src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java

```diff
@@ -149,48 +181,284 @@ protected TransPatterns(Context context) {
+                        indyType,
+                        staticArgValues);
+
+                JCFieldAccess qualifier = make.Select(make.QualIdent(bsm.owner), dynSym.name);
+                qualifier.sym = dynSym;
+                qualifier.type = syms.intType;
+                selector = make.Apply(List.nil(),
+                                      qualifier,
+                                      List.of(make.Ident(temp), make.Ident(index)))
+                               .setType(syms.intType);
             }
-            JCExpression translatedExpr = translate(tree.expr);
-            Type castTargetType = types.boxedTypeOrType(pattSym.erasure(types));
-
-            result = makeTypeTest(make.Ident(temp), make.Type(castTargetType));
-
-            VarSymbol bindingVar = bindingContext.bindingDeclared((BindingSymbol) patt.var.sym);
-            if (bindingVar != null) { //TODO: cannot be null here?
-                JCAssign fakeInit = (JCAssign)make.at(tree.pos).Assign(
-                        make.Ident(bindingVar), convert(make.Ident(temp), castTargetType)).setType(bindingVar.erasure(types));
-                LetExpr nestedLE = make.LetExpr(List.of(make.Exec(fakeInit)),
-                                                make.Literal(true));
-                nestedLE.needsCond = true;
-                nestedLE.setType(syms.booleanType);
-                result = makeBinary(Tag.AND, (JCExpression)result, nestedLE);
+
+            int i = 0;
+            boolean previousCompletesNormally = false;
+
+            for (var c : cases) {
+                List<JCCaseLabel> clearedPatterns = c.labels;
+                boolean hasJoinedNull =
+                        c.labels.size() > 1 && c.labels.stream().anyMatch(l -> l.isNullPattern());
+                if (hasJoinedNull) {
+                    clearedPatterns = c.labels.stream()
+                                              .filter(l -> !l.isNullPattern())
+                                              .collect(List.collector());
+                    if (clearedPatterns.head.hasTag(Tag.BINDINGPATTERN)) {
+                        ((JCBindingPattern) clearedPatterns.head).nullable = true;
+                    }
+                }
+                if (clearedPatterns.size() == 1 && clearedPatterns.head.isPattern() && !previousCompletesNormally) {
+                    JCCaseLabel p = clearedPatterns.head;
+                    bindingContext = new BasicBindingContext();
+                    VarSymbol prevCurrentValue = currentValue;
+                    try {
+                        currentValue = temp;
+                        JCExpression test = (JCExpression) this.<JCTree>translate(p);
+                        c.stats = translate(c.stats);
+                        JCContinue continueSwitch = make.Continue(null);
+                        continueSwitch.target = tree;
+                        c.stats = c.stats.prepend(make.If(makeUnary(Tag.NOT, test).setType(syms.booleanType),
```

Comment:
I assume this is code which "resumes" to the following `case` if the pattern doesn't match. I guess in most cases the bootstrap method would do the check anyway - but I suppose that with guards, the bootstrap method, alone, cannot guarantee the match. Also, it seems like this requires backend support for `continue` in switch. Again, all this needs to be better documented, it's pretty hard to infer all this by looking at the code.